### PR TITLE
Asterius rules

### PR DESCRIPTION
**Pull request by ylecornec**
This PR contains the following asterius related rules as well as the tests.

- The `ahc_dist` rule converts a haskell binary target (`haskell_test`, `haskell_binary` or `haskell_cabal_binary`) produced with the asterius haskell toolchain into an archive containing multiple javascript files.
There is a transition on its `dep` attribute so that this binary target is built using the asterius haskell toolchain.

- The `asterius_binary` and `asterius_test` rules can then run these files using node (if the `ahc_dist` target is built with `target="node"`).

- The `asterius_webpack` rule can bundle all of these file into a unique js file (if the `ahc_dist` target is built with `target="browser"`).

## Tests

The `tests/asterius/asterius_tests_utils.bzl` file contains a macro used for the asterius tests which we use to call the above rules on existing binary targets from the tests folder.

We call this macro for almost all the binaries which do not depend on C code or use GHC specific toolchains libraries.

There are three exceptions which need more investigation:

 - `tests/lhs` needs to call the `unlit` binary, which should maybe be made explicit in the haskell toolchain first.

 - `tests/haskell_cabal_package` imports its main function from `Lib`, and it is not visible as such in the asterius generated javascript files. This example also fails when calling `ahc-cabal` directly.

 - `tests/binary_with_main` is a bit similar because it makes use of the `main_function` attribute.

## Logs

I am unsure what to do about the logs of `ahc-build` (which always outputs some build info).
Such as :

```
[INFO] Writing JavaScript entry module to "/home/stan/.cache/bazel/_bazel_stan/3c7d084a53b480b3a5081e4bf80f09ce/sandbox/linux-sandbox/879/execroot/rules_haskell/bazel-out/k8-fastbuild/bin/tests/asterius/package-id-clash-binary/custom_output_name.mjs"
```

Should a flag be added to `ahc-build` in order to disable them?

### CODE REVIEW

File: haskell/asterius/defs.bzl [hunk 1]
```diff
@@ -31,3 +38,411 @@ asterius_toolchain = rule(
...
+
+    return [DefaultInfo(files = depset([output_file]))]
+
+asterius_webpack = rule(
+    asterius_webpack_impl,
+    attrs = {
+        "ahc_dist_dep": attr.label(
+            mandatory = True,
+            allow_single_file = True,
+            doc = """
```
Comment (aherrmann) [inline haskell/asterius/defs.bzl:330]:
```suggestion
            doc = """\
```
So that it's still the first line of the docstring? IIRC that has meaning in Startdoc.

File: haskell/asterius/defs.bzl [hunk 2]
```diff
@@ -31,3 +38,411 @@ asterius_toolchain = rule(
...
+    toolchains = [
+        "@rules_sh//sh/posix:toolchain_type",
+    ],
+    doc = "Creates a bundle using webpack out of ahc_dist target build for the browser.",
+)
+
+# Copied from https://github.com/tweag/asterius/blob/9f2574d9c2b50aa83d105741799e2f65b05e2023/asterius/test/ghc-testsuite.hs
+# The node options required to execute outputs from asterius.
+
+node_options = [
```
Comment (aherrmann) [inline haskell/asterius/defs.bzl:349]:
```suggestion
_node_options = [
```
Do these need to be public?

File: tests/haskell_cabal_library_sublibrary_name/package1.cabal [hunk 3]
```diff
@@ -7,8 +7,7 @@ executable package1_executable
     hs-source-dirs:   app
     default-language: Haskell2010
     build-depends:
-        Cabal
-        , base
+        base
```
Comment (aherrmann) [inline tests/haskell_cabal_library_sublibrary_name/package1.cabal:10]: What's the reason for this change?

**Review by aherrmann — COMMENTED** [review]
Great work! Really exciting to see so many tests pass already.

I left a couple comments / questions.

File: tests/package-reexport/Final.hs [hunk 4]
```diff
@@ -5,4 +5,4 @@ import qualified Dep
 import           Control.Exception (assert)
 
 main :: IO ()
-main = pure $ assert (TransitiveDep.whoAmI == Dep.whoIsMyDep) ()
+main = assert (TransitiveDep.whoAmI == Dep.whoIsMyDep) (pure ())
```
[the remark this answers is not in the document]
Comment (ylecornec) [reply tests/package-reexport/Final.hs:8]: It seems to be a laziness issue: `main = pure $ assert False ()` works fine, so the test could never fail (not only with asterius).

[the remark this answers is not in the document]
Comment (aherrmann) [reply tests/package-reexport/Final.hs:8]: Ah, interesting. That makes sense.

File: haskell/asterius/defs.bzl [hunk 5]
```diff
@@ -31,3 +38,411 @@ asterius_toolchain = rule(
...
+            values = ["browser", "node"],
+            doc = """
+            Whether the build output is intended to run with node, or on the browser.
+            The asterius_webpack rule only accepts dependencies built for the browser.
+            """,
+        ),
+        "main_output": attr.string(
+            mandatory = True,
+            doc = "The name for the output file corresponding to the entrypoint. It must terminate by '.mjs'",
+        ),
```
[the remark this answers is not in the document]
Comment (ylecornec) [reply haskell/asterius/defs.bzl:213]: I think it is useless in isolation. We need to know its name to pass it to the `nodejs_binary` via the `asterius_binary` macro, but maybe we could use a default one, and make this optional.

File: haskell/asterius/defs.bzl [hunk 6]
```diff
@@ -31,3 +38,411 @@ asterius_toolchain = rule(
...
+# to prevent conflicts as the library files have the same name.
+def _asterius_untar_impl(ctx):
+    posix_toolchain = ctx.toolchains["@rules_sh//sh/posix:toolchain_type"]
+    subfolder_name = ctx.attr.subfolder_name
+    ahc_dist_info = ctx.attr.dep[AhcDistInfo]
+
+    output_files = [
+        ctx.actions.declare_file(paths.join(subfolder_name, name))
+        for name in ahc_dist_info.files
+    ]
```
[the remark this answers is not in the document]
Comment (ylecornec) [reply haskell/asterius/defs.bzl:243]: This is because the generated library files all have the same names, so there were conflicts if two `ahc_dist` rules were in the same directory. But maybe there is a better way around it, like generating the files in a subfolder, and then use `pkg_tar`?

File: haskell/asterius/defs.bzl (hunk 5 above)
[the remark this answers is not in the document]
Comment (aherrmann) [reply haskell/asterius/defs.bzl:213]: Yes, the target name plus the extension might be a reasonable default.

File: haskell/asterius/defs.bzl (hunk 6 above)
[the remark this answers is not in the document]
Comment (aherrmann) [reply haskell/asterius/defs.bzl:243]: I see, thanks for explaining. Yes, I think a subfolder and emitting the files is a good solution to this. It avoids unecessary `tar/untar` operations. It also avoids leaving undeclared outputs in the execroot under `{workdir}`. These are not an issue for sandboxed builds, but with unsandboxed builds, like on Windows, these can cause issues.

File: tests/haskell_cabal_library_sublibrary_name/package1.cabal (hunk 3 above)
Comment (ylecornec) [reply tests/haskell_cabal_library_sublibrary_name/package1.cabal:10]: The `Cabal` library is not actually needed here and I thought that it was not compatible with asterius.
Because the asterius test failed with:  `Cabal is not a toolchain library`.

But now that I looked into it deeper I think it was a mistake.
We try to use Cabal as a toolchain library because it is in [this list](https://github.com/tweag/rules_haskell/blob/e67eff938b5c580663c1459d52abb26e90658524/haskell/cabal.bzl#L985-L1022), but this list seems wrong in the case of asterius. So maybe we could recover the Cabal package via some other way in this case ?

Comment (aherrmann) [reply tests/haskell_cabal_library_sublibrary_name/package1.cabal:10]: Oh, I see. Thanks for looking into this.
I think we can remove that list and instead use the location information provided by stack [here](https://github.com/tweag/rules_haskell/blob/e67eff938b5c580663c1459d52abb26e90658524/haskell/cabal.bzl#L1160-L1161). IIRC if there is no `location` field in the spec then it is a core package.
Though, I think this is orthogonal to this PR and perhaps better addressed in a separate PR.

**Comment by ylecornec** [comment]
Thank you, I addressed the various comments.

   - The `ahc_dist` rule now creates its output in a subfolder, so the `asterius_untar` rule was removed.
     If needed it is possible to use `pkg_tar` on the output but it is not asterius specific.
     There is an optional `subfolder_name` attribute to be able to control the name of this subfolder.
   - There is now a default value for the name of the `entry_point` file.
   - There is now a transition to configure the ahc_dist targets (for the browser or not).
      The [build setting](https://github.com/tweag/rules_haskell/pull/1621/files#diff-9d21ccabe299eeb98999e18b704133d701be776e90d771eb972751729b9878f2R160-R164) is declared in an external repository to work around the [following issue](https://github.com/bazelbuild/bazel/issues/12951) (This way the same name for this build setting can be used from inside or outside the rules_haskell repository).
   - The webpack rule used by asterius is now declared in an [external repository](https://github.com/tweag/rules_haskell/pull/1621/files#diff-9d21ccabe299eeb98999e18b704133d701be776e90d771eb972751729b9878f2R115-R130), which makes it possible for users to use a preexisting `rules_nodejs` setup to override it. Alongside `asterius_dependencies_bindist` and `asteriuis_dependencies_nix`, there is a now a third rule for this purpose:
     `asterius_dependencies_custom` which uses the label of an custom webpack rule. For instance defined as such:

        ```
        load("@npm//webpack-cli:index.bzl", webpack = "webpack_cli")
        webpack(
            name = "webpack",
            visibility = ["//visibility:public"],
        )
        ```
   - The asterius test with the Cabal library dependency is removed for now.
   - The asterius test with the java dependency as well.
   - There is now a genrule in the test macro, that tries to calls the `asterius_binary` target.

File: haskell/asterius/defs.bzl [hunk 7]
```diff
@@ -31,3 +38,411 @@ asterius_toolchain = rule(
...
+
+    (output_prefix, _) = paths.split_extension(ctx.attr.main_output)
+    output_tar_file = ctx.actions.declare_file("{}.tar".format(ctx.label.name))
+    for ext in [".wasm", ".wasm.mjs", ".req.mjs"]:
+        all_output_files.append(output_prefix + ext)
+
+    # ctx.file.dep was generated in the asterius platform configuration,
+    # and we want to generate js files in the current configuration.
+    # So we will copy it to the folder corresponding to the current platform.
+    file_copy_path = paths.join(output_tar_file.dirname, ctx.file.dep.basename)
```
[the remark this answers is not in the document]
Comment (ylecornec) [reply haskell/asterius/defs.bzl]: This should be fine now that we output files in a subfolder.
At the moment, it is not possible to set `--platforms @rules_haskell//haskell/asterius:asterius_platform` globally when compiling a `ahc_dist` target because it depends on the `rules_nodejs` toolchain that sets a target constraint (and maybe should not). But only the dependencies of this target need to be built for the asterius_platform.

**Review by aherrmann — APPROVED** [review]
Thank you! Great work!

**Comment by aherrmann** [comment]
CI is failing after rebase. Perhaps due to the changes in https://github.com/tweag/rules_haskell/pull/1623.
